lib/charms/grafana_k8s/v0/grafana_dashboard.py

```diff
@@ -0,0 +1,938 @@
+                :method:`ops.model.Relation.id`.
+
+        Returns: a list of known dashboards coming from the provided relation instance.
+        """
+        return [
+            self._to_external_object(relation_id, dashboard)
+            for dashboard in self._stored.dashboards.get(relation_id, [])
+        ]
+
+    def _on_grafana_dashboard_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle relation changes in related providers.
+
+        If there are changes in relations between Grafana dashboard consumers
+        and providers, this event handler (if the unit is the leader) will
+        get data for an incoming grafana-dashboard relation through a
+        :class:`GrafanaDashboardsChanged` event, and make the relation data
+        available in the app's datastore object. The Grafana charm can
+        then respond to the event to update its configuration.
+        """
+        # TODO Are we sure this is right? It sounds like every Grafana unit
+        # should create files with the dashboards in its container.
+        if not self._charm.unit.is_leader():
+            return
+
+        self._render_dashboards_and_emit_event(event.relation)
+
+    def update_dashboards(self, relation: Optional[Relation] = None) -> None:
+        """Re-establish dashboards on one or more relations.
+
+        If something changes between this library and a datasource, try to re-establish
+        invalid dashboards and invalidate active ones.
+
+        Args:
+            relation: a specific relation for which the dashboards have to be
+                updated. If not specified, all relations managed by this
+                :class:`GrafanaDashboardConsumer` will be updated.
+        """
+        if not self._charm.unit.is_leader():
+            return
+
+        relations = [relation] if relation else self._charm.model.relations[self._relation_name]
+
+        for relation in relations:
+            self._render_dashboards_and_emit_event(relation)
+
+    def _on_grafana_dashboard_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Update job config when providers depart.
+
+        When a Grafana dashboard provider departs, the configuration
+        for that provider is removed from the list of dashboards
+        """
+        if not self._charm.unit.is_leader():
+            return
+
+        self._remove_all_dashboards_for_relation(event.relation)
+
+    def _render_dashboards_and_emit_event(self, relation: Relation) -> None:
+        """Validate a given dashboard.
+
+        Verify that the passed dashboard data is able to be found in our list
+        of datasources and will render. If they do, let the charm know by
+        emitting an event.
+
+        Args:
+            relation: Relation; The relation the dashboard is associated with.
+        """
+        other_app = relation.app
+
+        if not (raw_data := relation.data[other_app].get("dashboards", {})):
```

Comment:
```suggestion
        raw_data = relation.data[other_app].get("dashboards", {})
        if not raw_data:
```
Do we need this library to support Python 3.5. If so we may need to remove the assignment expression (walrus).